pkg/trafficrouting/manager.go

```diff
@@ -215,16 +221,39 @@ func (m *Manager) FinalisingTrafficRouting(c *TrafficRoutingContext, onlyRestore
 		klog.Errorf("%s newTrafficRoutingController failed: %s", c.Key, err.Error())
 		return false, err
 	}
-
+	noCanaryService := c.OnlyTrafficRouting || c.DisableGenerateCanaryService
+	// The "already-finalised" conditions of this FinalisingTrafficRouting function are:
+	// 1. the stable service has no selector
+	// 2. AND canary service has been cleaned up
+	var stableServiceRestored, canaryServiceRemoved bool
+	// check condition 1
+	stableService := &corev1.Service{ObjectMeta: metav1.ObjectMeta{Namespace: c.Namespace, Name: trafficRouting.Service}}
+	if err := m.Get(context.TODO(), client.ObjectKeyFromObject(stableService), stableService); err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("%s get stable service(%s) failed: %s", c.Key, trafficRouting.Service, err.Error())
+		return false, err
+	} else {
+		stableServiceRestored = errors.IsNotFound(err) || stableService.Spec.Selector[c.RevisionLabelKey] == ""
+	}
+	// check condition 2
 	cService := &corev1.Service{ObjectMeta: metav1.ObjectMeta{Namespace: c.Namespace, Name: cServiceName}}
-	// if canary svc has been already cleaned up, just return
-	// even DisableGenerateCanaryService is true, canary svc still exists, because canary service is stable service
-	if err = m.Get(context.TODO(), client.ObjectKeyFromObject(cService), cService); err != nil {
-		if !errors.IsNotFound(err) {
-			klog.Errorf("%s get canary service(%s) failed: %s", c.Key, cServiceName, err.Error())
-			return false, err
-		}
-		// In rollout failure case, no canary-service will be created, this step ensures that the canary-ingress can be deleted in a time.
+	if err := m.Get(context.TODO(), client.ObjectKeyFromObject(cService), cService); err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("%s get canary service(%s) failed: %s", c.Key, cServiceName, err.Error())
+		return false, err
+	} else {
+		// if noCanaryService is true, we have never created canary service
+		canaryServiceRemoved = errors.IsNotFound(err) || noCanaryService
+	}
+	// only if both conditions are met
+	if canaryServiceRemoved && stableServiceRestored {
+		/*
+			even both of the conditions are met, we call Finalise
```

Comment:
> even both of the conditions are met

 if both of the conditions are met

Reply:
the logic here is bloated and complex, i will simplify it in the coming pr